compiler/rustc_codegen_ssa/src/mir/block.rs

```diff
@@ -732,6 +731,109 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     }
                 };
 
+                if intrinsic == sym::const_eval_select {
+                    let place = match &args[2] {
+                        mir::Operand::Copy(place) | mir::Operand::Move(place) => Some(place),
+                        mir::Operand::Constant(_) => None,
+                    };
+                    if let Some(place) = place {
+                        let ty = self.mir.local_decls[place.local].ty;
+                        match ty.kind() {
+                            ty::FnDef(..) | ty::FnPtr(..) => {}
+                            ty::Closure(_, substs) => {
```

Comment:
Is this even reachable? AFAIK `Closure` disappears long before MIR reaches codegen.

Why does the interpreter only need a `FnDef` but this needs more cases?

Reply:
Because the ctfe interpreter does not need to handle closures? Does `ty::Closure` really disappear?

Reply:
Mir never sees `ty::Closure`. So I assume neither does codegen. Closures become `FnDef` calls by the time they reach this stage of compilation.

There is still `FnPtr` support in codegen though and not in CTFE. That should also be made consistent - either by removing `FnPtr` support in codegen, or by implementing it in both of them *and ensuring we have tests for it*.

Reply:
> Closures become `FnDef` calls by the time they reach this stage of compilation.

But the problem here is that the MIR is passing a closure as an argument, and not actually calling it. So I guess there is no MIR transformation for that?

Reply:
But even then, there is no good reason for this to differ between codegen and CTFE, I think. Also having to do a bunch of complicated extra stuff for closures is somewhat surprising. There is probably another API that avoids this problem.

For example, I think we could require both arguments to be `FnDef`, and then implement a wrapper like
```rust
const_eval_select_inner<I, C, R, O>(
  arg: I,
  arg_const: C,
  arg_runtime: R,
  fn_const: impl FnOnce(I, C) -> O, // must be a FnDef type (but we don't have a trait for that)
  fn_runtime: impl FnOnce(I, R) -> O, // must be a FnDef type (but we don't have a trait for that)
) -> R;

fn const_eval_select<I, F_C: FnOnce(I) -> O, F_R: impl FnOnce(I) -> O, O>(arg: I, c: F_C, r: F_R) -> O {
  fn call_it<I, F: FnOnce(I) -> O, O>(arg: I, f: F) -> O { f(arg) }

  const_eval_select_inner(
    arg,
    c,
    r,
    call_it::<I, F_C, O>,
    call_it::<I, F_R, O>,
  )
}
```

Reply:
That won't fix #100696 and would just be equivalent to the current code. I'm going to remove the logic around closures and see what I can do.

Reply:
Oh I see.

But if slicing directly used `const_eval_select_inner`, then that could fix https://github.com/rust-lang/rust/issues/100696, couldn't it?